### ISSUE TITLE
How to speed-up calls into rust

### ISSUE THREAD

**Issue by fabianmurariu** [bytecodealliance/wasmtime#9686]
Project [here](https://github.com/fabianmurariu/jit-experiments)

I'm evaluationg wasm + wasmtime for a query execution library, much of that involves calling iterators into rust to get data then executing dynamically generated logic via wasm.

The code here sums up an iterator by calling a rust function and doing the summing up in wasm

1. is there a better way to measure the wasmtime execution time
2. can this be made to run faster? If yes then how?

currently the run times seem to be

```
rust: 30ns
wasm: 21ms
python: 0.75s
```

**Comment by alexcrichton** [comment]
At least for wasmtime you're measuring what's the fastest way. The Rust number probably isn't representative because it's probably constant-folded by LLVM which means it's not actually doing anything at runtime. The Wasm number shows an overhead of 4.2ns per call which is roughly what's expected. I'm not familiar with benchmarking Python so I don't know whether that's representative.

**Comment by alexcrichton** [comment]
Yes, I'd recommend reading over the [examples here](https://docs.rs/wasmtime/latest/wasmtime/struct.Func.html#examples-2)

**Comment by alexcrichton** [comment]
I'm going to close this as the discussion has died down and we're not aware of otherwise lower-hanging fruit to pick here in terms of optimizing. If such fruit is found though it's definitely something we're interested to help with!
